### PR TITLE
feat(server): add Gradle build insights and cache analytics

### PR DESCRIPTION
**Pull request by fortmarek**
## Summary

Adds server-side infrastructure and dashboard pages for **Gradle build insights and cache analytics**.

This is the server foundation that the Tuist Gradle plugin will report build data to, enabling teams to understand their Gradle cache performance at the task level.

image

image

image

image

### Why this design?

Gradle, Bazel, and Xcode/Tuist have **fundamentally different cache architectures**:

| System | Levels | Storage Format | Deduplication |
|--------|--------|---------------|---------------|
| **Bazel** | 2 (Action Cache + CAS) | Individual files by digest | Yes, file-level |
| **Xcode/Tuist** | 2 (Cacheable Task + CAS) | Individual files by node_id | Yes, file-level |
| **Gradle** | 1 (Task → tar.gz) | Single archive per task | No |

Xcode/Tuist and Bazel use a **two-level CAS model** where tasks map to content-addressable storage entries (individual `.o`, `.swiftmodule` files). Multiple tasks producing identical files share CAS entries, enabling file-level deduplication.

Gradle uses a **single-level model** via `TarBuildCacheEntryPacker` — one cache key maps to one atomic tar.gz blob containing all task outputs. There is no sub-artifact tracking or CAS equivalent.

**Implication**: The Gradle implementation is simpler than Xcode. We don't need a `gradle_cas_outputs` table — just task-level tracking with build-level aggregated outcome counts.

### Gradle build model

Gradle builds are **task-centric**. Each build executes multiple tasks (e.g., `:app:compileKotlin`, `:app:assembleDebug`), and each task has an **outcome**:

| Outcome | Description | Cacheable? |
|---------|-------------|------------|
| `FROM_CACHE` | Output restored from build cache | Yes (cache hit) |
| `UP_TO_DATE` | No work needed (incremental) | N/A (local avoidance) |
| `EXECUTED` | Task ran and produced output | Yes (cache miss) |
| `FAILED` | Task execution failed | No |
| `SKIPPED` | Task was skipped | No |
| `NO_SOURCE` | No input sources found | No |

### Key metrics

- **Cache hit rate**: `FROM_CACHE / cacheable_tasks_count` — percentage of cacheable tasks served from cache
- **Avoidance rate**: `(FROM_CACHE + UP_TO_DATE) / total_tasks` — total work avoided
- **Task outcome breakdown**: Distribution across all 6 outcome types
- **Cache event analytics**: Upload/download volume and counts with trend calculations
- **Percentile analytics**: P50/P90/P99 cache hit rates with time-series data

### Data flow

```
Gradle Build (Tuist Gradle Plugin)
        │
        │  Build completes → POST /api/v1/projects/:id/gradle/builds
        ▼
┌─────────────────────────────────────────────────────────────────┐
│  Build Report (JSON)                                            │
│  - gradle_version, java_version, status, git metadata           │
│  - tasks: [{path, type, outcome, cacheable, duration, ...}]     │
└────────────────────┬────────────────────────────────────────────┘
                     │
        ┌────────────┴────────────┐
        ▼                         ▼
┌───────────────────┐    ┌────────────────────┐
│  Server API       │    │  Cache Service     │
│  POST /gradle/    │    │  GET/PUT /gradle/  │
│  builds           │    │  :cache_key        │
└────────┬──────────┘    └────────┬───────────┘
         │                        │
         ▼                        ▼
┌─────────────────────────────────────────────┐
│              ClickHouse                     │
│  ┌─────────────────┐  ┌──────────────────┐  │
│  │ gradle_builds   │  │ gradle_tasks     │  │
│  │ (build summary  │  │ (task-level      │  │
│  │  + aggregated   │  │  outcomes)       │  │
│  │  task counts)   │  │                  │  │
│  └─────────────────┘  └──────────────────┘  │
│  ┌──────────────────────────────────────┐   │
│  │ gradle_cache_events                  │   │
│  │ (upload/download from cache service) │   │
│  └──────────────────────────────────────┘   │
└─────────────────────┬───────────────────────┘
                      │
                      ▼
           ┌─────────────────────┐
           │  Dashboard          │
           │  (GradleInsightsLive)│
           │                     │
           │  • Cache Hit Rate   │
           │  • Avoidance Rate   │
           │  • Task Outcomes    │
           │  • Cache Transfers  │
           └─────────────────────┘
```

All Gradle data lives in **ClickHouse** (analytics-focused), unlike Xcode which uses PostgreSQL for `build_runs` (operational) + ClickHouse for CAS events.

### Why not an OpenAPI-generated client?

Unlike the Swift CLI which uses an OpenAPI-generated client, the Gradle plugin uses a hand-rolled HTTP layer with `java.net.HttpURLConnection`. This is intentional:

- **Minimal API surface** — only 2 endpoints (cache GET/PUT and build report POST), so code generation adds more complexity than it saves
- **Classpath conflict avoidance** — Gradle plugins share a classpath with other plugins and the build script. Adding dependencies like OkHttp or Retrofit risks version conflicts with the user's project. While this can be solved with a [Shadow JAR](https://github.com/GradleUp/shadow) (fat JAR with relocated packages), that adds build complexity we don't need yet for just 2 endpoints.
- **Cache endpoint isn't typical REST** — raw `application/octet-stream` binary transfer with custom auth retry logic, which OpenAPI generators don't model well
- **Auth is delegated to the CLI** — the plugin shells out to `tuist cache config` for token management, keeping the HTTP layer intentionally thin

Worth revisiting with a Shadow JAR setup if the API surface grows significantly.

### Example build scan

For comparison, here's a [Gradle Build Scan](https://scans.gradle.com/s/b4ylflapkzy3k) generated from the `examples/gradle/simple_android_app/` example project.

### Database (ClickHouse migrations)
- `gradle_builds` — Build-level data with aggregated task outcome counts, partitioned by month with 90-day TTL
- `gradle_tasks` — Task-level outcome tracking (path, type, outcome, cacheable, duration, cache_key)
- `gradle_cache_events` — Upload/download tracking from the cache service

### Server modules
- `Tuist.Gradle` — Context module for build/task/cache event CRUD
- `Tuist.Gradle.Analytics` — All analytics queries (cache hit rate, avoidance rate, percentiles, trends, time-series, task breakdown, cache events)
- `Tuist.Gradle.{Build, Task, CacheEvent}` — ClickHouse Ecto schemas

### API endpoints
- `POST /api/v1/projects/:id/gradle/builds` — Create build with task data
- `GET /api/v1/projects/:id/gradle/builds` — List builds (with pagination and date filtering)
- `GET /api/v1/projects/:id/gradle/builds/:id` — Get build details with tasks
- `GET /api/v1/projects/:id/gradle/analytics` — Combined analytics (all metrics in parallel)

### Dashboard
- `GradleInsightsLive` — LiveView dashboard with cache hit rate (avg + percentiles), avoidance rate, task outcome breakdown, cache transfer stats, and time-series trend charts
- Navigation entry in the project sidebar

### Authorization
- `gradle` object with `:create` and `:read` actions following existing patterns
- Supports user roles, project tokens, and account tokens with `project:gradle:read`/`project:gradle:write` scopes

### Cache service pipeline
- Telemetry handlers for Gradle cache download/upload events
- Broadway pipeline (`GradleCacheEventsPipeline`) mirroring the existing CAS events pattern
- Webhook endpoint (`/webhooks/gradle-cache`) with HMAC signature verification

### Tests
- 17 analytics tests covering all metrics, edge cases (zero data), multi-build aggregation, and combined parallel queries
- Cache pipeline and webhook controller tests

## Test plan

- [x] All 17 analytics tests pass (`mix test test/tuist/gradle/analytics_test.exs`)
- [x] Credo passes with no issues
- [x] Code formatted
- [x] End-to-end test with Gradle plugin reporting a build (see below)
- [ ] Run full test suite to verify no regressions
- [ ] Verify ClickHouse migrations apply cleanly in staging
- [ ] Manual dashboard verification with seeded data

<details>
<summary>E2E test script used during development</summary>

```bash
#!/usr/bin/env bash
set -euo pipefail

TUIST_BIN="/Users/marekfort/.local/share/mise/installs/tuist/4.138.1/tuist"
SERVER_URL="http://localhost:8080"
PROJECT_DIR="$(cd "$(dirname "$0")" && pwd)"
FULL_HANDLE="tuist/android-app"

cd "$PROJECT_DIR"
eval "$(mise env)"

clean_caches() {
    rm -rf ~/.gradle/caches/build-cache-*
    rm -rf "$PROJECT_DIR/.gradle" "$PROJECT_DIR/app/build" "$PROJECT_DIR/build"
}

run_build() {
    local label="$1"
    echo ""
    echo "=========================================="
    echo " $label"
    echo "=========================================="
    gradle --stop 2>/dev/null || true
    TUIST_EXECUTABLE="$TUIST_BIN" TUIST_SERVER_URL="$SERVER_URL" gradle build --info 2>&1 \
        | grep -iE "Could not load|Stored cache|remote build cache|Tuist:|FROM-CACHE|status 40" \
        || true
    echo ""
}

check_api() {
    echo "=========================================="
    echo " Build reports from API"
    echo "=========================================="
    local token
    token=$(cd /tmp && "$TUIST_BIN" cache config "$FULL_HANDLE" --json --url "$SERVER_URL" 2>/dev/null \
        | python3 -c "import sys,json; print(json.load(sys.stdin)['token'])")
    curl -s "$SERVER_URL/api/projects/$FULL_HANDLE/gradle/builds?limit=3" \
        -H "Authorization: Bearer $token" \
        | python3 -c "
import sys, json
data = json.load(sys.stdin)
for b in data['builds']:
    print(f\"Build {b['id'][:13]}...  local_hit={b['tasks_local_hit_count']}  remote_hit={b['tasks_remote_hit_count']}  executed={b['tasks_executed_count']}  up_to_date={b['tasks_up_to_date_count']}  hit_rate={b['cache_hit_rate']}%\")
"
}

echo "=== E2E Test: local_hit / remote_hit tracking ==="

# Build 1: cold build — everything executes, pushes to remote cache
clean_caches
run_build "Build 1: Cold build (should execute + store to remote)"

# Build 2: clear local cache — should pull from remote cache (remote_hit)
clean_caches
run_build "Build 2: After clearing local cache (should show remote_hit)"

# Check results via API
check_api
```

</details>

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: gradle/src/main/kotlin/dev/tuist/gradle/TuistHttpClient.kt [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+class TokenExpiredException : Exception()
+
+class TuistHttpClient(
+    private val configurationProvider: ConfigurationProvider,
+    private val connectTimeoutMs: Int = 30_000,
+    private val readTimeoutMs: Int = 60_000
+) {
+    @Volatile
+    private var cachedConfig: TuistCacheConfiguration? = null
```
Comment (fortmarek) [inline gradle/src/main/kotlin/dev/tuist/gradle/TuistHttpClient.kt:14]: `@Volatile` ensures that reads/writes to `cachedConfig` are visible across threads immediately — without it, a thread could see a stale value from its CPU cache. This is needed because Gradle runs tasks in parallel, and the double-checked locking pattern below relies on the field being volatile to work correctly.

File: server/lib/tuist/projects/project.ex [hunk 2]
```diff
@@ -33,6 +33,7 @@ defmodule Tuist.Projects.Project do
     field :report_days_of_week, {:array, :integer}, default: []
     field :report_schedule_time, :utc_datetime
     field :report_timezone, :string
+    field :build_system, Ecto.Enum, values: [xcode: 0, gradle: 1], default: :xcode
```
Comment (fortmarek) [inline server/lib/tuist/projects/project.ex:36]: we could consider supporting both build systems for a single project but it will make things quite complicated in pages like overview. I don't see much of a benefit of doing that right now, so I'd start with 1 project == 1 build system

File: server/lib/tuist_web/router.ex [hunk 3]
```diff
@@ -381,6 +381,12 @@ defmodule TuistWeb.Router do
           post "/", BuildsController, :create
         end
 
+        scope "/gradle" do
+          post "/builds", GradleController, :create_build
+          get "/builds", GradleController, :list_builds
+          get "/builds/:build_id", GradleController, :get_build
+        end
```
Comment (fortmarek) [inline server/lib/tuist_web/router.ex:388]: I'm keeping the URLs for the dashboard without `gradle` in the prefix as we want to keep the dashboard structure the same. In the API, I think it's better to be explicit about the build system.

**Comment by fortmarek** [comment]
@asmitbm The Gradle cache empty state (`server/lib/tuist_web/live/gradle_cache_live.html.heex`, lines 14 & 17) currently reuses the Xcode cache images (`empty_xcode_cache_light.png` / `empty_xcode_cache_dark.png`). Could you create Gradle-specific empty state images to replace them?

File: server/lib/tuist/gradle/build.ex [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    belongs_to :built_by_account, Tuist.Accounts.Account, foreign_key: :account_id, define_field: false
+    field :duration_ms, Ch, type: "UInt64"
+    field :gradle_version, Ch, type: "Nullable(String)"
+    field :java_version, Ch, type: "Nullable(String)"
+    field :is_ci, Ch, type: "Bool"
+    field :status, Ch, type: "Enum8('success' = 0, 'failure' = 1, 'cancelled' = 2)"
+    field :git_branch, Ch, type: "Nullable(String)"
+    field :git_commit_sha, Ch, type: "Nullable(String)"
+    field :git_ref, Ch, type: "Nullable(String)"
+    field :root_project_name, Ch, type: "Nullable(String)"
```
Comment (cschmatzler) [inline server/lib/tuist/gradle/build.ex]: these should probably be non-Nullable `String` with default `""` as per https://clickhouse.com/docs/optimize/avoid-nullable-columns

File: server/lib/tuist/gradle.ex [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+      tasks_up_to_date_count: task_counts.up_to_date,
+      tasks_executed_count: task_counts.executed,
+      tasks_failed_count: task_counts.failed,
+      tasks_skipped_count: task_counts.skipped,
+      tasks_no_source_count: task_counts.no_source,
+      cacheable_tasks_count: task_counts.cacheable,
+      inserted_at: now
+    }
+
+    IngestRepo.insert_all(Build, [build_entry])
```
Comment (cschmatzler) [inline server/lib/tuist/gradle.ex]: not a requirement right now, but would make a note somewhere to pull this into the `IngestBuffer` strategy once we get some traffic on this.

File: server/lib/tuist/gradle/build.ex
Comment (cschmatzler) [inline server/lib/tuist/gradle/build.ex:1]: Think at some point we need to do a bit where we reorganize modules, e.g. `Tuist.Builds.Build` needs to become `Tuist.Xcode.Build` to match this, and be explicit and consistent about build systems

Comment (fortmarek) [reply server/lib/tuist/gradle/build.ex:1]: yes, totally. For now, I think it's fine Xcode is considered the "default", but at some point, we'll need to reorganize things, so both Gradle and Xcode are on the same footing.

File: cache/test/cache/gradle_cache_events_pipeline_test.exs [hunk 6]
```diff
@@ -0,0 +1,137 @@
+defmodule Cache.GradleCacheEventsPipelineTest do
+  use ExUnit.Case, async: false
```
Comment (pepicrft) [inline cache/test/cache/gradle_cache_events_pipeline_test.exs]: What about using [this recommended approach](https://hexdocs.pm/broadway/Broadway.html#module-testing) to be able to run the tests of this test module in parallel?

File: cache/test/cache/gradle_cache_events_pipeline_test.exs [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+            acknowledger: {Broadway.CallerAcknowledger, {self(), make_ref()}, :ok}
+          }
+        end)
+
+      expect(Req, :request, fn options ->
+        assert options[:url] == "http://localhost:4000/webhooks/gradle-cache"
+        assert options[:method] == :post
+
+        body = options[:body]
+        decoded_body = Jason.decode!(body)
```
Comment (pepicrft) [inline cache/test/cache/gradle_cache_events_pipeline_test.exs]: Let's use JSON here, and I'd add that to `AGENTS.md` and hope that coding agents will pick it up?

File: cache/test/cache/gradle_cache_events_pipeline_test.exs [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+        assert Enum.at(decoded_body["events"], 0)["project_handle"] == project_handle
+        assert Enum.at(decoded_body["events"], 0)["action"] == "upload"
+        assert Enum.at(decoded_body["events"], 0)["size"] == 1024
+        assert Enum.at(decoded_body["events"], 0)["cache_key"] == "gradle-cache-key-123"
+
+        assert Enum.at(decoded_body["events"], 1)["account_handle"] == account_handle
+        assert Enum.at(decoded_body["events"], 1)["project_handle"] == project_handle
+        assert Enum.at(decoded_body["events"], 1)["action"] == "download"
+        assert Enum.at(decoded_body["events"], 1)["size"] == 2048
+        assert Enum.at(decoded_body["events"], 1)["cache_key"] == "gradle-cache-key-456"
```
Comment (pepicrft) [inline cache/test/cache/gradle_cache_events_pipeline_test.exs]: Is there a reason why we are not comparing against a list? Is it because there are some item attributes you don't want to match against?

File: cache/test/cache_web/controllers/cas_controller_test.exs [hunk 9]
```diff
@@ -1,5 +1,5 @@
 defmodule CacheWeb.CASControllerTest do
-  use CacheWeb.ConnCase, async: true
+  use CacheWeb.ConnCase, async: false
```
Comment (pepicrft) [inline cache/test/cache_web/controllers/cas_controller_test.exs:5]: Why?

File: gradle/src/main/kotlin/dev/tuist/gradle/TuistBuildCache.kt [hunk 10]
```diff
@@ -217,11 +231,9 @@ class TuistBuildCacheService(
     override fun store(key: BuildCacheKey, writer: BuildCacheEntryWriter) {
         if (!isPushEnabled) return
 
-        val config = getOrRefreshConfig() ?: return
-        val url = buildCacheUrl(config, key.hashCode)
-
-        executeWithRetry<Unit> { currentConfig ->
-            val connection = openConnection(url, currentConfig)
+        httpClient.execute<Unit> { config ->
```
Comment (pepicrft) [inline gradle/src/main/kotlin/dev/tuist/gradle/TuistBuildCache.kt:234]: Why did we remove the retries here in the case of token expired errors?

File: gradle/src/main/kotlin/dev/tuist/gradle/TuistPlugin.kt [hunk 11]
```diff
@@ -85,17 +121,29 @@ class TuistPlugin : Plugin<Settings> {
...
+     * The base URL that points to the Tuist server. Defaults to https://tuist.dev.
+     */
+    var url: String = "https://tuist.dev"
+
+    /**
+     * Whether to upload build insights in the background. When null (default),
+     * the upload runs in the background for local builds and in the foreground on CI
+     * to ensure it completes before ephemeral agents exit.
+     */
+    var uploadInBackground: Boolean? = null
```
Comment (pepicrft) [inline gradle/src/main/kotlin/dev/tuist/gradle/TuistPlugin.kt:145]: Why `Boolean?` and not `Boolean`?

File: gradle/src/main/kotlin/dev/tuist/gradle/TuistBuildInsights.kt [hunk 12]
```diff
@@ -0,0 +1,460 @@
...
+            gradleVersion = parameters.gradleVersion.orNull,
+            rootProjectName = parameters.rootProjectName.orNull,
+            ciDetector = ciDetector,
+            gitInfoProvider = gitInfoProvider
+        )
+
+        val baseUrl = parameters.url.get().trimEnd('/')
+
+        val response = httpClient.execute { config ->
+            val url = URI("$baseUrl/api/projects/$accountHandle/$projectHandle/gradle/builds")
```
Comment (pepicrft) [inline gradle/src/main/kotlin/dev/tuist/gradle/TuistBuildInsights.kt]:
```suggestion
            val url = URI($baseUrl).resolve("/api/projects/$accountHandle/$projectHandle/gradle/builds")
```

File: server/lib/tuist/projects.ex [hunk 13]
```diff
@@ -270,7 +270,8 @@ defmodule Tuist.Projects do
       name: name,
       account_id: account_id,
       created_at: created_at,
-      visibility: visibility
+      visibility: visibility,
+      build_system: Keyword.get(opts, :build_system, :xcode)
```
Comment (pepicrft) [inline server/lib/tuist/projects.ex:274]: We have `:xcode` as the default fallback in several places. I'd probably extract that inside `Tuist.Projects.Project.default_build_system()` and call it from here.

**Comment by asmitbm** [comment]
Empty state page for Gradle cache: https://www.figma.com/design/bq2VJ91tEamZI7snIXPHNc/Prototypes?node-id=6191-163662&t=nkemrJVNsUuuOpuN-4
image

**Comment by asmitbm** [comment]
Select build system on create project page: https://www.figma.com/design/bq2VJ91tEamZI7snIXPHNc/Prototypes?node-id=5310-297516&t=nkemrJVNsUuuOpuN-4
image

**Comment by asmitbm** [comment]
Project overview with badges: https://www.figma.com/design/bq2VJ91tEamZI7snIXPHNc/Prototypes?node-id=5310-309659&t=nkemrJVNsUuuOpuN-4

image

File: gradle/src/main/kotlin/dev/tuist/gradle/TuistPlugin.kt (hunk 11 above)
Comment (fortmarek) [reply gradle/src/main/kotlin/dev/tuist/gradle/TuistPlugin.kt:145]: Because when not specified, it's resolved dynamically based on whether the environment is ci or not

File: cache/test/cache/gradle_cache_events_pipeline_test.exs (hunk 7 above)
Comment (fortmarek) [reply cache/test/cache/gradle_cache_events_pipeline_test.exs]: > Let's use JSON here, and I'd add that to AGENTS.md and hope that coding agents will pick it up?

Let's not treat `AGENTS.md` as dumpground. This should be a credo rule.

File: server/lib/tuist/projects.ex (hunk 13 above)
Comment (fortmarek) [reply server/lib/tuist/projects.ex:274]: > We have :xcode as the default fallback in several places. I'd probably extract that inside Tuist.Projects.Project.default_build_system() and call it from here.

I think instead we should remove the default once we force teams to pick a build system when creating a project instead

File: cache/test/cache_web/controllers/cas_controller_test.exs (hunk 9 above)
Comment (fortmarek) [reply cache/test/cache_web/controllers/cas_controller_test.exs:5]: The CASEventsPipeline Broadway processor runs in a shared GenServer process. When it processes an event triggered by the controller tests, it calls `Cache.Authentication.server_url()`. Mimic stubs are scoped to the calling process by default — with `async: true`, the Broadway process can't see the test process's stubs, causing a timeout on `S3TransfersBuffer.flush()`. Setting `async: false` makes the stubs globally visible so the Broadway processor can resolve them.

File: gradle/src/main/kotlin/dev/tuist/gradle/TuistBuildCache.kt (hunk 10 above)
Comment (fortmarek) [reply gradle/src/main/kotlin/dev/tuist/gradle/TuistBuildCache.kt:234]: The retries are still there — they were moved from `TuistBuildCacheService.executeWithRetry()` into `TuistHttpClient.execute()` (line 26-44 in TuistHttpClient.kt). Both `load` and `store` now call `httpClient.execute { ... }` which catches `TokenExpiredException`, refreshes the config, and retries. This deduplicates the retry logic so it's shared across all HTTP operations (build cache + build insights).

File: cache/test/cache/gradle_cache_events_pipeline_test.exs (hunk 6 above)
Comment (fortmarek) [reply cache/test/cache/gradle_cache_events_pipeline_test.exs]: Changed to `async: true` — these tests call `handle_batch/4` directly in the test process, so they don't need global stub visibility. Also removed the `handle_message` test and setup block that had the global stub.

File: cache/test/cache/gradle_cache_events_pipeline_test.exs (hunk 7 above)
Comment (fortmarek) [reply cache/test/cache/gradle_cache_events_pipeline_test.exs]: Updated to use `Jason.decode!` for the JSON body comparison.

File: cache/test/cache/gradle_cache_events_pipeline_test.exs (hunk 8 above)
Comment (fortmarek) [reply cache/test/cache/gradle_cache_events_pipeline_test.exs]: Updated to compare against the full list directly.

File: gradle/src/main/kotlin/dev/tuist/gradle/TuistBuildInsights.kt (hunk 12 above)
Comment (fortmarek) [reply gradle/src/main/kotlin/dev/tuist/gradle/TuistBuildInsights.kt]: Updated to use `URI.resolve()` 👍
